addon/-private/system/relationships/state/has-many.js

```diff
@@ -111,6 +111,32 @@ export default class ManyRelationship extends Relationship {
     super.removeCanonicalInternalModelFromOwn(internalModel, idx);
   }
 
+  unloadInverseInternalModelFromOwn(internalModel) {
+    super.unloadInverseInternalModelFromOwn(internalModel);
+
+    const canonicalIndex = this.canonicalState.indexOf(internalModel);
+
+    if (canonicalIndex !== -1) {
+      this.canonicalState.splice(canonicalIndex, 1);
+    }
+
+    const manyArray = this._manyArray;
+
+    if (manyArray) {
+      const currentState = manyArray.currentState;
+      const idx = currentState.indexOf(internalModel);
+
+      if (idx !== -1) {
+        currentState.splice(idx, 1);
+
+        manyArray.set('length', currentState.length);
+        manyArray.notifyPropertyChange('length');
```

Comment:
for `async` relationships an issue was surfaced in which `set` was not enough to notify the `PromiseProxy` that `length` had changed. This particular notify was a shot in the dark at fixing it that did not work and should be removed. See comment in `model.js` for more on the issue surfaced, you may have an idea as to what isn't wired correctly.